Guarantee mapping output for all variants

bencap opened the pull request:
### Enhancements to error handling:
* Added detailed error messages for cases where accession values cannot be determined during allele or haplotype annotation, replacing generic `ValueError` exceptions (`src/dcd_mapping/annotate.py`). [[1]](diffhunk://#diff-80cb2a97f07b0bfc40eb19abe98b007a9d85f34652197e19a30b1096a28f0041L269-R277) [[2]](diffhunk://#diff-80cb2a97f07b0bfc40eb19abe98b007a9d85f34652197e19a30b1096a28f0041L324-R335)
* Improved error messages for transcript selection failures, including specific reasons for the failure, such as missing gene symbols or matching transcripts (`src/dcd_mapping/transcripts.py`). [[1]](diffhunk://#diff-fb6a30da3e1820798884d5f9fc8322774e0ebf872f1a86bec47a848eb258b815L57-R60) [[2]](diffhunk://#diff-fb6a30da3e1820798884d5f9fc8322774e0ebf872f1a86bec47a848eb258b815L182-R186)
* Added error handling for unsupported sequence namespaces and intronic variants in genomic mapping, returning `MappedScore` objects with appropriate error messages (`src/dcd_mapping/vrs_map.py`). [[1]](diffhunk://#diff-0295e435516bca854a6f27ebee70141e4b90633b060ee294381b07b195941795L456-R483) [[2]](diffhunk://#diff-0295e435516bca854a6f27ebee70141e4b90633b060ee294381b07b195941795L612-R642)

### Improvements to mapping logic:
* Introduced checks for mismatches between the number of mapped scores and total score records, with corresponding error messages (`src/api/routers/map.py`).

### Codebase simplification:
* Removed hard-coded logic for specific datasets, such as manually selecting protein annotations or predefined results for certain score sets (`src/dcd_mapping/annotate.py`, `src/dcd_mapping/transcripts.py`). [[1]](diffhunk://#diff-80cb2a97f07b0bfc40eb19abe98b007a9d85f34652197e19a30b1096a28f0041L522-L526) [[2]](diffhunk://#diff-fb6a30da3e1820798884d5f9fc8322774e0ebf872f1a86bec47a848eb258b815L326-L338)
* Simplified conditional statements by directly assigning error messages without redundant checks (`src/dcd_mapping/annotate.py`). [[1]](diffhunk://#diff-80cb2a97f07b0bfc40eb19abe98b007a9d85f34652197e19a30b1096a28f0041L298-R302) [[2]](diffhunk://#diff-80cb2a97f07b0bfc40eb19abe98b007a9d85f34652197e19a30b1096a28f0041L354-R361)

### Performance and timeout adjustments:
* Increased the HTTP request timeout from 30 to 60 seconds in the `http_download` function to accommodate slower network conditions (`src/dcd_mapping/resource_utils.py`).

bencap committed:
Bump dataset file download timeout to 60 seconds

This should help slightly with timeout issues we see in production

sallybg committed:
Change function name to patch_target_sequence_type

sallybg committed:
Bump version number

This update changes how alignment is performed for some score sets,
so bump major version.

bencap committed:
Enforce a 1 to 1 relationship between incoming score rows and outgoing annotated mappings

Prior to this change, it was possible for some score rows to generate valid mappings with other score rows not creating a mapped variant. This had some negative downstream consequences, which will be remedied by ensuring that if any variant receives a mapped variant, all variants receive a mapped variant.

bencap committed:
fix: vrs_ref_allele_seq extension could potentially have an empty string as its sequence

bencap committed:
Bump dev Seqrepo version

bencap committed:
fix: update logic in patch_target_sequence_type to handle cases with no nucleotide variants

sallybg commented on src/dcd_mapping/vrs_map.py:
```diff
@@ -609,8 +612,11 @@ def _map_genomic(
+        return MappedScore(
+            accession_id=row.accession,
+            score=row.score,
+            error_message=f"Reference sequence namespace not supported: {namespace}",
```
I wonder if we should consider ending the whole job if this happens, because an error with a reference sequence is unexpected and will affect multiple variants.

sallybg commented on src/dcd_mapping/vrs_map.py:
```diff
@@ -783,7 +789,14 @@ def _map_accession(
+            MappedScore(
+                accession_id=row.accession,
+                score=row.score,
+                error_message="Could not generate mapped allele objects. No sequence id was provided.",
```
Same comment as above, I wonder if this should be a mapping job-wide error rather than returning individual variant mapping errors.

sallybg commented on src/dcd_mapping/vrs_map.py:
```diff
@@ -802,8 +815,14 @@ def _map_accession(
+            MappedScore(
+                accession_id=row.accession,
+                score=row.score,
+                error_message=f"Unrecognized accession prefix for accession id {metadata.target_accession_id}",
```
Same as above about returning individual errors vs scoreset-wide errors.

sallybg reviewed (commented):
This looks good! There are a few places where we are now returning individual variant-level errors when there is a problem with a reference sequence. I lean toward having the whole mapping job fail if these errors happen, although I recognize that for multi-target score sets we'd then be failing all targets if even one target fails.

bencap committed:
feat: add custom exceptions for VRS mapping errors and update error handling in mapping functions

bencap committed:
feat: refactor custom exceptions to exceptions module

bencap committed:
Merge branch 'mavedb-dev' into feature/bencap/45/guarantee-mapping-output-for-all-variants

bencap commented:
Thanks @sallybg, totally agree. 

I added some custom exceptions for those possibilities and included them in the try/except block during mapping. I also went ahead and refactored the other custom exceptions we have into a dedicated exception module to keep them a little more organized.

sallybg reviewed (approved):
Looks great!
